### DISCUSSION TITLE
Need help in FAST API Starlette Request. Please find the errors:

### DISCUSSION THREAD

**Discussion by pypi-ahmad** [fastapi/fastapi#9931] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Error at auth :

'Traceback (most recent call last):\n  File "c:\\Users\\.vscode\\extensions\\ms-python.python-2023.12.0\\pythonFiles\\lib\\python\\debugpy\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_resolver.py", line 189, in _get_py_dictionary\n    attr = getattr(var, name)\n  File " \\env\\lib\\site-packages\\starlette\\requests.py", line 155, in auth\n    ), "AuthenticationMiddleware must be installed to access request.auth"\nAssertionError: AuthenticationMiddleware must be installed to access request.auth\n'


Error at Session
'Traceback (most recent call last):\n  File "c:\\Users\\.vscode\\extensions\\ms-python.python-2023.12.0\\pythonFiles\\lib\\python\\debugpy\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_resolver.py", line 189, in _get_py_dictionary\n    attr = getattr(var, name)\n  File " \env\\lib\\site-packages\\starlette\\requests.py", line 148, in session\n    ), "SessionMiddleware must be installed to access request.session"\nAssertionError: SessionMiddleware must be installed to access request.session\n'
```

### Description

The uvicorn app is created using fastapi. The app is running okay, but when trying to send any request locally, it is showing an error "HTTP/1.1 500 Internal Server Error", It seems like the error occurs while send authentication request that return nothing due to authorizatio failure.

This is how our request looks like :
![image](https://github.com/tiangolo/fastapi/assets/72138709/932c03a5-3774-4865-8bed-754826c0d307)

Error at auth :

'Traceback (most recent call last):\n  File "c:\\Users\\.vscode\\extensions\\ms-python.python-2023.12.0\\pythonFiles\\lib\\python\\debugpy\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_resolver.py", line 189, in _get_py_dictionary\n    attr = getattr(var, name)\n  File " [\\env\\lib\\site-packages\\starlette\\requests.py](file://env/lib/site-packages/starlette/requests.py)", line 155, in auth\n    ), "AuthenticationMiddleware must be installed to access request.auth"\nAssertionError: AuthenticationMiddleware must be installed to access request.auth\n'

Error at Session
'Traceback (most recent call last):\n  File "c:\\Users\\.vscode\\extensions\\ms-python.python-2023.12.0\\pythonFiles\\lib\\python\\debugpy\\_vendored\\pydevd\\_pydevd_bundle\\pydevd_resolver.py", line 189, in _get_py_dictionary\n    attr = getattr(var, name)\n  File " \env\\lib\\site-packages\\starlette\\requests.py", line 148, in session\n    ), "SessionMiddleware must be installed to access request.session"\nAssertionError: SessionMiddleware must be installed to access request.session\n'

This is uvicorn application running :
![image](https://github.com/tiangolo/fastapi/assets/72138709/5003275f-31e5-417a-bfbd-d1ccfae94bfc)
![image](https://github.com/tiangolo/fastapi/assets/72138709/64e74036-2566-4a00-89f7-4fa7796622e0)

We installed startlette as well - pip install starlette-auth-toolkit
Not sure why Authorization Failure is happening. Any help is greatly appreciated.

### Operating System

Windows

### FastAPI Version

0.89.0

### Python Version

Python 3.9.1

**Comment by jgould22** [comment, marked as answer]
As the error message says you need to install the SessionMiddleware and AuthenticationMiddleware

It is going to look something like 

```python
from fastapi import FastAPI
from starlette.middleware.sessions import SessionMiddleware
from starlette.middleware.authentication import AuthenticationMiddleware

app = FastAPI()
# More options here https://www.starlette.io/middleware/#sessionmiddleware
app.add_middleware(SessionMiddleware, secret_key="some_key")
# More here https://www.starlette.io/authentication/
app.add_middleware(AuthenticationMiddleware, backend=SomeBackend)
```

**Reply by pypi-ahmad** [reply]
Shall I install this in main.py file?

**Reply by jgould22** [reply]
You need to add the middleware to your application instance, if that is in your main.py then you can add it there.

**Reply by pypi-ahmad** [reply]
We are using this piece of code

"""

from fastapi import FastAPI, Request

app = FastAPI()

globals.initialize_app_globals()
globals.initialize_logger()
obj_logging_app_insights = globals.obj_logging_app_insights


@app.middleware("http")
async def initialize_app(req: Request, next):
        try:

        if 'X-Arr-Log-Id' in req.headers:
            globals.LOGID = req.headers['X-Arr-Log-Id']
       
        
        globals.obj_logging_app_insights.logger.info(f"{globals.LOGID} - Request came to the fastapi-middleware")
        # with obj_logging_app_insights.tracer.span("main") as span:
        # 
        #     span.span_kind = SpanKind.SERVER
        response = await next(req)
       
        return response

   except ConnectionError as e:
        obj_logging_app_insights.set_custom_property({constants.SQL_STATUS_CODE: constants.SQL_STATUS_MESSAGE})
        obj_logging_app_insights.log_custom_property_to_app_insights(e)
        res = response.create_response(status_code = constants.SQL_STATUS_CODE)
        return res
    except  Exception as e:
        res = response.create_response(status_code = constants.GENERIC_EXCEPTION_STATUS_CODE)
        obj_logging_app_insights.set_custom_property({constants.GENERIC_EXCEPTION_STATUS_CODE : constants.GENERIC_EXCEPTION_STATUS_MESSAGE})
        obj_logging_app_insights.log_custom_property_to_app_insights(e)
        return res
"""

We are using decorators, this @app.middleware("http")

while getting request, there is some issue with authentication middleware and session middleware.

![image](https://github.com/tiangolo/fastapi/assets/72138709/9a02dc18-0d55-4e6e-b08f-5307bfb3b104)

**Reply by jgould22** [reply]
The decorators are for defining your own middleware.

You need to use middleware from starlette with app.add_middleware as shown below.

```python
# More options here https://www.starlette.io/middleware/#sessionmiddleware
app.add_middleware(SessionMiddleware, secret_key="some_key")
# More here https://www.starlette.io/authentication/
app.add_middleware(AuthenticationMiddleware, backend=SomeBackend)
```
also note: `SomeBackend` is not a real backend you will have to define this yourself.  Check this link for more information https://www.starlette.io/authentication/
